sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs

```diff
@@ -22,5 +24,183 @@ public partial class LogsIngestionClient
         protected LogsIngestionClient()
         {
         }
+
+        // The size we use to determine whether to upload as a single PUT BLOB
+        // request or stage as multiple blocks.
+        private const int _singleUploadThreshold = 1000000; // 1 Mb in byte format
+
+        internal HttpMessage CreateUploadRequest(string ruleId, string streamName, RequestContent content, string contentEncoding, RequestContext context)
+        {
+            var message = _pipeline.CreateMessage(context, ResponseClassifier204);
+            var request = message.Request;
+            request.Method = RequestMethod.Post;
+            var uri = new RawRequestUriBuilder();
+            uri.Reset(_endpoint);
+            uri.AppendPath("/dataCollectionRules/", false);
+            uri.AppendPath(ruleId, true);
+            uri.AppendPath("/streams/", false);
+            uri.AppendPath(streamName, true);
+            uri.AppendQuery("api-version", _apiVersion, true);
+            request.Uri = uri;
+            request.Headers.Add("Accept", "application/json");
+            request.Headers.Add("Content-Type", "application/json");
+            if (contentEncoding != null)
+            {
+                request.Headers.Add("Content-Encoding", contentEncoding);
+            }
+            if (contentEncoding == "gzip")
+            {
+                GZipUtf8JsonRequestContent gzContent = new(content);
+                request.Content = gzContent;
+            }
+            else
+            {
+                request.Content = content;
+            }
+            return message;
+        }
+
+        /// <summary>
+        /// Hidden method for batching data
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="logEntries"></param>
+        /// <returns></returns>
+        internal static IEnumerable<BinaryData> Batching<T>(IEnumerable<T> logEntries)
+        {
+            //TODO: use Array pool instead
+            MemoryStream stream = new MemoryStream(_singleUploadThreshold);
+            WriteMemory(stream, BinaryData.FromString("[").ToMemory());
+            int entryCount = 0;
+            foreach (var log in logEntries)
+            {
+                BinaryData entry = log is BinaryData d ? d : BinaryData.FromObjectAsJson(log);
+
+                var memory = entry.ToMemory();
+                if ((stream.Length + memory.Length + 1) >= _singleUploadThreshold) // if adding this entry makes stream > 1 Mb send current stream now
+                {
+                    WriteMemory(stream, BinaryData.FromString("]").ToMemory());
+                    stream.Position = 0; // set Position to 0 to return everything from beginning of stream
+                    yield return (BinaryData.FromStream(stream));
```

Comment:
We should probably just return the stream itself to avoid an extra memory copy into BinaryData.